### PR TITLE
[Merged by Bors] - Reflection cleanup

### LINKED ISSUE #1528: 'bevy_render::pipeline::state_descriptors::PrimitiveTopology\' does not support ReflectValue serialization
## Bevy version

bc4fe9b186db3f32eef38fc4241289df480fa949

## Operating system & version

Windows 10

## What you did

spawn a PbrBundle and try to serialize it like in the example (https://github.com/bevyengine/bevy/blob/main/examples/scene/scene.rs)

## What you expected to happen

get a serialized RON string of my scene

## What actually happened

My game crashed and produced the following error message:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { code: Message("Type \'bevy_render::pipeline::state_descriptors::PrimitiveTopology\' does not support ReflectValue serialization"), position: Position { line: 0, col: 0 } }', src\shared\gameplay\scene_loading.rs:9:57
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

## Additional information

Discord user Bd#4705 on Bevy discord server said:

> It's missing #[reflect_value(Serialize, Deserialize, PartialEq, Hash)], all reflected enums should have it. There are a few places where the reflection of bevy types is not complete, you should probably open a broader issue for that.

### PR DESCRIPTION
**Pull request by CleanCut**
This is an effort to provide the correct `#[reflect_value(...)]` attributes where they are needed.  

Supersedes #1533 and resolves #1528.

---

I am working under the following assumptions (thanks to @bjorn3 and @Davier for advice here):

- Any `enum` that derives `Reflect` and one or more of { `Serialize`, `Deserialize`, `PartialEq`, `Hash` } needs a `#[reflect_value(...)]` attribute containing the same subset of { `Serialize`, `Deserialize`, `PartialEq`, `Hash` } that is present on the derive.
- Same as above for `struct` and `#[reflect(...)]`, respectively.
- If a `struct` is used as a component, it should also have `#[reflect(Component)]`
- All reflected types should be registered in their plugins

I treated the following as components (added `#[reflect(Component)]` if necessary):
- `bevy_render`
  - `struct RenderLayers`
- `bevy_transform`
  - `struct GlobalTransform`
  - `struct Parent`
  - `struct Transform`
- `bevy_ui`
  - `struct Style`

Not treated as components:
- `bevy_math`
  - `struct Size<T>`
  - `struct Rect<T>`
  - Note: The updates for `Size<T>` and `Rect<T>` in `bevy::math::geometry` required using @Davier's suggestion to add `+ PartialEq` to the trait bound. I then registered the specific types used over in `bevy_ui` such as `Size<Val>`, etc. in `bevy_ui`'s plugin, since `bevy::math` does not contain a plugin.
- `bevy_render`
  - `struct Color`
  - `struct PipelineSpecialization`
  - `struct ShaderSpecialization`
  - `enum PrimitiveTopology`
  - `enum IndexFormat`

Not Addressed:
- I am not searching for components in Bevy that are _not_ reflected. So if there are components that are not reflected that should be reflected, that will need to be figured out in another PR.
- I only added `#[reflect(...)]` or `#[reflect_value(...)]` entries for the set of four traits { `Serialize`, `Deserialize`, `PartialEq`, `Hash` } _if they were derived via `#[derive(...)]`_. I did not look for manual trait implementations of the same set of four, nor did I consider any traits outside the four.  Are those other possibilities something that needs to be looked into?

### CODE REVIEW

**Comment by Davier** [comment]
`#[reflect_value(...)]` should not be used for structs, they need `#[reflect(...)]` instead. I have no idea what happens when using both attributes on the same type, but I don't think it's supposed to happen.

`Size<T>` seems to work by adding the `T: PartialEq` bound, but I'll look at the macro to see if it can be avoided.

Other things needed:
- check that all reflected components have `#[reflect(Component)]`
- check that all reflected types are registered in their plugins
- (probably out of scope of this PR) check that all components are reflected

**Comment by CleanCut** [comment]
@Davier Thanks!  That's exactly the sort of thing I need to know and wasn't obvious from skimming the source.  I've updated my assumptions list in the original comment, and I'll go back and update the PR accordingly.



Question:

**Comment by CleanCut** [comment]
Ready for another review!

@Davier Did you find anything we should do with `Size<T: Reflect = f32>` or `Rect<T: Reflect>` in this PR?

**Comment by Davier** [comment]
> @Davier Did you find anything we should do with `Size<T: Reflect = f32>` or `Rect<T: Reflect>` in this PR?

So the issue is that `#[derive(PartialEq)]` implements `PartialEq` only for `T: PartialEq`, and `#[derive(Reflect)]` implements `Reflect` only for `T: Reflect`. Hence when the `Reflect` impl defers the comparison to `Size::eq` with `#[reflect(PartialEq)]`, it can't find it.
I think it's fine do `struct Size<T: Reflect + PartialEq = f32>`, but @cart might want to weight in.

Everything else looks good to me.

**Comment by CleanCut** [comment]
@Davier Awesome. Thanks to your advice, I was able to add `#[reflect(PartialEq)]` to both `Size<T>` and `Rect<T>` after changing the full bound to `T: Reflect + PartialEq = f32`.  I then registered the specific usages over in `bevy_ui`, since `bevy_ui` has a plugin (`bevy_math` does not) and `bevy_ui` is the only place that uses them.

Sidenote: what is `= f32` in the trait bounds? I have never seen that trailing equal sign before in a trait bound, I don't know what it means, and I am having trouble finding documentation about it. Do you know where I can read up about that to understand what is going on?

**Comment by CleanCut** [comment]
@cart Ready for review.

File: crates/bevy_sprite/src/sprite.rs [hunk 1]
```diff
@@ -9,7 +9,6 @@ use bevy_render::{
     texture::Texture,
 };
 use serde::{Deserialize, Serialize};
-
```
Comment (Davier) [inline crates/bevy_sprite/src/sprite.rs:12]: You probably didn't mean to remove that line?

**Review by Davier — APPROVED** [review]
Everything looks in order :-)

File: crates/bevy_sprite/src/sprite.rs (hunk 1 above)
Comment (CleanCut) [reply crates/bevy_sprite/src/sprite.rs]: Correct. ✅ Fixed

**Comment by Davier** [comment]
> Sidenote: what is `= f32` in the trait bounds? I have never seen that trailing equal sign before in a trait bound, I don't know what it means, and I am having trouble finding documentation about it. Do you know where I can read up about that to understand what is going on?

It's a default type for the generic, such that `Size` is equivalent to `Size<f32>`. There seems to be a real lack of documentation on that though. I only found [this RFC](https://github.com/rust-lang/rfcs/blob/master/text/0213-defaulted-type-params.md), and the [Reference](https://doc.rust-lang.org/reference/items/generics.html) doesn't explain them either.

**Comment by CleanCut** [comment]
@Davier Thanks!!!  That makes so much sense. I can't believe I've gone so many years without encountering this!

**Comment by CleanCut** [comment]
I went ahead and merged the work @Davier helped me with [to implement GetTypeRegistrations for types that manually implement Reflect](https://github.com/CleanCut/bevy/pull/2) into this pull request.

**Comment by cart** [comment]
Looks good to me. Thanks!

**Pull request closed without merging**